### DISCUSSION TITLE
How to clean up user-defined ASN1 methods?

### DISCUSSION THREAD

**Discussion by gingerbeard-man** [openssl/openssl#23587] [Q&A] [answered]
I have an application that creates and registers half a dozen ASN1_METHODs (all copies of the default EC method but with bespoke ids).  Valgrind tells me these are "still reachable".

It seems there is no way to pop entries back of the app_methods stack.  Indeed the stack itself is allocated on first use but never freed.

Is this intentional?  Am I missing some obvious way to appease Valgrind?

(Error checking removed for brevity)

```
EVP_PKEY_ASN1_METHOD *meth_new = NULL;
const EVP_PKEY_ASN1_METHOD *meth_default = EVP_PKEY_asn1_find(NULL, NID_X9_62_id_ecPublicKey);

meth_new = EVP_PKEY_asn1_new(OBJ_txt2nid(OBJ_G2_ECC_WITH_SHA256), 0, "EC", "G2 ECC with SHA256");
EVP_PKEY_asn1_copy(meth_new, meth_default);
EVP_PKEY_asn1_add0(meth_new);

meth_new = EVP_PKEY_asn1_new(OBJ_txt2nid(OBJ_G2_ECC_WITH_SHA384), 0, "EC", "G2 ECC with SHA384");
EVP_PKEY_asn1_copy(meth_new, meth_default);
EVP_PKEY_asn1_add0(meth_new);

meth_new = EVP_PKEY_asn1_new(OBJ_txt2nid(OBJ_G2_ECC_WITH_SHA512), 0, "EC", "G2 ECC with SHA512");
EVP_PKEY_asn1_copy(meth_new, meth_default);
EVP_PKEY_asn1_add0(meth_new);

...
```

If I were to keep each meth_new as a separate variable then I could free them later (using EVP_PKEY_asn1_find to recover them is messy due to its const return).  But this seems dangerous as the freed memory will still be referenced by the entries on the app_methods stack.  Also the stack itself is never freed.

https://github.com/openssl/openssl/blob/85cf92f55d9e2ac5aacf92bedd33fb890b9f8b4c/crypto/asn1/ameth_lib.c#L24
https://github.com/openssl/openssl/blob/85cf92f55d9e2ac5aacf92bedd33fb890b9f8b4c/crypto/asn1/ameth_lib.c#L162-L166

**Comment by t8m** [comment, marked as answer]
Yes, there is no way to free these method objects. As these EVP_PKEY_asn1 functions are deprecated I do not think it makes sense to add an explicit  cleanup function for them. I assume as a bug fix it would be acceptable to have a patch to clean them in OPENSSL_cleanup().
